packages/plugin-ext/src/hosted/browser/hosted-plugin.ts

```diff
@@ -534,18 +535,37 @@ export class HostedPluginSupport {
     }
 
     protected createServerRpc(pluginHostId: string): RPCProtocol {
-        const emitter = new Emitter<string>();
+
+        const onCloseEmitter = new Emitter<ChannelCloseEvent>();
+        const onErrorEmitter = new Emitter<unknown>();
+        const onMessageEmitter = new Emitter<MessageProvider>();
+
+        // Create RPC protocol before adding the listener to the watcher to receive the watcher's cached messages after the rpc protocol was created.
+        const rpc = new RPCProtocolImpl({
+            close: () => {
+                onCloseEmitter.dispose();
+                onErrorEmitter.dispose();
+                onMessageEmitter.dispose();
+            },
+            getWriteBuffer: () => {
+                const writer = new Uint8ArrayWriteBuffer();
+                writer.onCommit(buffer => {
+                    this.server.onMessage(pluginHostId, buffer);
+                });
+                return writer;
+            },
+            onClose: onCloseEmitter.event,
+            onError: onErrorEmitter.event,
+            onMessage: onMessageEmitter.event
+        });
```

Comment:
This could deserve a dedicated function. What do you think?

The same logic is also in `packages/plugin-ext/src/hosted/browser/plugin-worker.ts` and `packages/plugin-ext/src/hosted/browser/worker/worker-main.ts`.

Reply:
Introduced a new "BasicChannel" class with  https://github.com/eclipse-theia/theia/commit/2b4016e8da50466f271bce2b2f808bf0ed371a10 to reduce the need for code duplication.